pip/req/req_file.py

```diff
@@ -97,166 +70,152 @@ def parse_requirements(filename, finder=None, comes_from=None, options=None,
-                    value = relative_to_reqs_file
-                finder.find_links.append(value)
-
-        # ---------------------------------------------------------------------
-        elif linetype == IGNORE:
-            pass
-
-
-def parse_line(line):
-    if not line.startswith('-'):
-        # Split the requirement from the options.
-        if ' --' in line:
-            req, opts = line.split(' --', 1)
-            opts = parse_requirement_options('--%s' % opts)
-        else:
-            req = line
-            opts = {}
-
-        return REQUIREMENT, (req, opts)
-
-    firstword, rest = partition_line(line)
-    # -------------------------------------------------------------------------
-    if firstword == '-e' or firstword == '--editable':
-        return REQUIREMENT_EDITABLE, rest
-
-    # -------------------------------------------------------------------------
-    if firstword == '-r' or firstword == '--requirement':
-        return REQUIREMENT_FILE, rest
-
-    # -------------------------------------------------------------------------
-    if firstword in parser_flags:
-        if rest:
-            msg = 'Option %r does not accept values.' % firstword
-            raise RequirementsFileParseError(msg)
-        return FLAG, firstword
+def process_line(line, filename, line_number, finder=None, comes_from=None,
+                 options=None, session=None, cache_root=None):
+    """
+    Process a single requirements line; This can result in creating/yielding
+    requirements, or updating the finder.
+    """
 
-    # -------------------------------------------------------------------------
-    if firstword in parser_options:
-        if not rest:
-            msg = 'Option %r requires value.' % firstword
-            raise RequirementsFileParseError(msg)
-        return OPTION, (firstword, rest)
+    parser = build_parser()
+    args = shlex.split(line)
+    opts, args = parser.parse_args(args)
+    req = None
+
+    if args:
+        # don't allow multiple requirements
+        if len(args) > 1:
+            msg = 'Only one requirement supported per line.'
+            raise ReqFileOnlyOneReqPerLineError(msg)
+        for key, value in opts.__dict__.items():
+            # only certain options can be on req lines
+            dest_strings = [o().dest for o in SUPPORTED_OPTIONS_REQ]
+            if value is not None and key not in dest_strings:
+                # get the option string
+                # the option must be supported to get to this point
+                for o in SUPPORTED_OPTIONS:
+                    o = o()
+                    if o.dest == key:
+                        opt_string = o.get_opt_string()
+                msg = ('Option not supported on a'
+                       ' requirement line: %s' % opt_string)
+                raise ReqFileOptionNotAllowedWithReqError(msg)
+        req = args[0]
+
+    # don't allow multiple/different options (on non-req lines)
+    if not args and len(
```

Comment:
I would split this into two statements to improve readability:

```python
values = [v for v in opts.__dict__.values() if v is not None]
if not args and len(values) > 1:
```

I don't mean to dwell on it, but this is an  example of how the 80 column line limit is hurting readability in pip's case. The splitting of the 'Option not supported on a required line: %' above is another.